docs/DafnyRef/Refinement.md

```diff
@@ -92,32 +92,133 @@ TODO
 
 * the parent has no initialization,
 * the child has the same type as the parent, and
 * one or both of the following holds:
    * the child has an initializing expression
-   * the child is declared `ghost` and the parent is not `ghost`, or vice versa
+   * the child is declared `ghost` and the parent is not `ghost`, or vice versa.
 
-To indicate it is a refining declaration, a refining `const` declaration
-contains an ellipsis in this syntax:
-```grammar
-"const" { Attribute } CIdentType "..." [ ":=" Expression ]
-```
+A refining module can also introduce new `const` declarations that do
+not exist in the refinement parent.
 
 ## 21.5. Method declarations
 
-TODO
+A child `method` definition may refine a parent `method` declaration or
+definition to
+
+* provide a body missing in the parent,
+* replace the body of the parent with a new, semantically-compatible
+  body,
+* strengthen the postcondition of the parent method by adding one or more
+  `ensures` clauses,
+* provide a more deterministic version of a non-deterministic parent
+  body, or
+* amend the body of the parent method with additional statements.
+
+The type signature of a refining method must be the same as that of the
+parent method it refines. This can be ensured by providing an explicit
+type signature equivalent to that of the parent (with renaming of
+parameters allowed) or by using an ellipsis (`...`) to indicate copying
+of the parent type signature. The following example illustrates that
+latter case.
+
+```
+module M0 {
+  method m(x: int) returns (r: int)
+}
+
+module M1 refines M0 {
+  method m...
+  {
+    return x + 1;
+  }
+}
+```
+
+As an example of semantic compatibility and increased determinism,
+consider the following example.
+
+```
+module M0 {
+  method m(x: int) returns (r: int)
+    ensures r > x;
+  {
+    var y :| y > x;
+    return y;
+  }
+}
+
+module M1 refines M0 {
+  method m(x: int) returns (r: int)
+  {
+    return x + 1;
+  }
+}
+```
+
+Here, `m` ensures that its result will be greater than `x`, but the
+version in `M0` returns an arbitrary value satisfying that
+postcondition. The version of `m` in `M1` is deterministic, and is
+required to satisfy all of the postconditions that its parent method
+declares.
+
+To introduce additional statements, the child method can include
+ellipses within the body to stand in for portions of code from the
+parent body. Dafny then attempts to merge the body of the child with the
+body of the parent by filling in the ellipses. For example:
+
+```
+module M0 {
+  method m(x: int) returns (r: int)
+    ensures r > x;
+  {
+    var y :| y > x;
+    return y;
+  }
+}
+
+module M1 refines M0 {
+  method m(x: int) returns (r: int)
+  {
+    ...;
+    return y + 1;
+  }
+}
+```
+
+Refinement of method bodies can go far beyond adding statements to the
+end, however. Full details of the algorithm used to perform the merge
+operation are available in [this
+paper](https://dl.acm.org/doi/10.1007/s00165-012-0254-3).
```

Comment:
For anyone interested in these details, the long comment before class `RefinementTransformer` in `Source/Dafny/RefinementTransformer.cs` may also be useful.

Also, it would be good to point out that hover text in the IDE will tell you what statements were "inherited" from the parent. To try this, hover over a `...;` statement, over a `}` (before which there is an implicit `...;`), or some other statement that has been changed.

Reply:
> To try this, hover over a ...; statement, over a } (before which there is an implicit ...;), or some other statement that has been changed.

Sweet